pkg/apis/capabilities/v1beta1/openapi_types.go

```diff
@@ -0,0 +1,193 @@
+package v1beta1
+
+import (
+	"reflect"
+
+	"github.com/3scale/3scale-operator/pkg/common"
+
+	"github.com/go-logr/logr"
+	"github.com/google/go-cmp/cmp"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+)
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+const (
+	OpenAPIKind = "OpenAPI"
+
+	// OpenAPIInvalidConditionType represents that the combination of configuration
+	// in the OpenAPISpec is not supported. This is not a transient error, but
+	// indicates a state that must be fixed before progress can be made.
+	// Example: the spec references invalid openapi spec
+	OpenAPIInvalidConditionType common.ConditionType = "Invalid"
+
+	// OpenAPIReadyConditionType indicates the openapi resource has been successfully reconciled.
+	// Steady state
+	OpenAPIReadyConditionType common.ConditionType = "Ready"
+
+	// OpenAPIFailedConditionType indicates that an error occurred during reconcilliation.
+	// The operator will retry.
+	OpenAPIFailedConditionType common.ConditionType = "Failed"
+)
+
+// OpenAPIRefSpec Reference to the OpenAPI Specification
+type OpenAPIRefSpec struct {
+	// ConfigMapRef ConfigMap that contains the OpenAPI Document
+	// +optional
+	ConfigMapRef *corev1.ObjectReference `json:"configMapRef,omitempty"`
+
+	// URL Remote URL from where to fetch the OpenAPI Document
+	// +kubebuilder:validation:Pattern=`^https?:\/\/.*$`
+	// +optional
+	URL *string `json:"url,omitempty"`
+}
+
+// OpenAPISpec defines the desired state of OpenAPI
+type OpenAPISpec struct {
+	// OpenAPIRef Reference to the OpenAPI Specification
+	OpenAPIRef OpenAPIRefSpec `json:"openapiRef"`
+
+	// ProviderAccountRef references account provider credentials
+	// +optional
+	ProviderAccountRef *corev1.LocalObjectReference `json:"providerAccountRef,omitempty"`
+
+	// ProductionPublicBaseURL Custom public production URL
+	// +kubebuilder:validation:Pattern=`^https?:\/\/.*$`
+	// +optional
+	ProductionPublicBaseURL string `json:"productionPublicBaseURL,omitempty"`
```

Comment:
Is this parameter optional? If so, I think it might be better to define it as a pointer to string to follow Kubernetes conventions https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required.

I'm aware that in this case (and probably others) the zero-value is "forbidden" because it might imply it is unset itself, but they still insist on using pointers (see some of the arguments in the API conventions guide for that too).
I also think would also be more consistent with what we are doing in other places.

Same would apply to other attributes I've seen of this CR.